Optimize github action cache

tequdev opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

- ccache 
dev branch: keep all caches (as existing) 
other: cache only differences from cache of dev branch

- conan 
dev branch: keep all cache (as existing) 
other: use dev branch cache and keep cache only if conanfile is changed

- nix/macos workflow 
Run workflow at 0:00 UTC daily so that cache is always present

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

tequdev committed:
optimize github action cache

tequdev committed:
fix

sublimator commented on .github/actions/xahau-ga-dependencies/action.yml:
```diff
@@ -42,6 +42,26 @@ runs:
+      run: |
+        # Check if we're on the main branch
+        if [ "${{ github.ref_name }}" == "${{ inputs.main_branch }}" ]; then
+          echo "conanfile-changed=true" >> $GITHUB_OUTPUT
```
pedantry: maybe save it as should-save-conan-cache or similar, as the name doesn't actually reflect whether the conanfile-changed

sublimator commented on .github/actions/xahau-ga-build/action.yml:
```diff
@@ -76,6 +88,22 @@ runs:
+        
+        if [ "${{ steps.safe-branch.outputs.name }}" == "${{ inputs.main_branch }}" ]; then
+          # default branch: use default branch cache
+          ccache --set-config=cache_dir="$HOME/.ccache-main"
```
pedantry:

Maybe this should go to the existing
.github/actions/xahau-configure-ccache/action.yml

sublimator replied:
https://github.com/tequdev/xahaud/pull/3

sublimator commented on .github/actions/xahau-ga-build/action.yml:
```diff
@@ -48,12 +48,24 @@ runs:
+        restore-keys: |
+          ${{ runner.os }}-ccache-v${{ inputs.cache_version }}-${{ inputs.compiler-id }}-${{ inputs.configuration }}-
+          ${{ runner.os }}-ccache-v${{ inputs.cache_version }}-${{ inputs.compiler-id }}-
+          ${{ runner.os }}-ccache-v${{ inputs.cache_version }}-
```
This last restore key, without a compiler discriminant is probably not very useful

sublimator replied:
also in https://github.com/tequdev/xahaud/pull/3 if you want it

sublimator committed:
refactor: improve github actions cache optimization (#3)

- move ccache configuration logic to dedicated action
- rename conanfile-changed to should-save-conan-cache for clarity

sublimator commented:
Related #559 also touched this

tequdev committed:
Merge branch 'dev' into cache-optimize

tequdev committed:
Merge branch 'dev' into cache-optimize

RichardAH committed:
Merge branch 'dev' into cache-optimize
